File: walk_route.js
// walk_route.js
#!/usr/bin/env node
const fs = require('fs');
const axios = require('axios');
const { program } = require('commander');

const DATA_FILE = 'walk_route.json';
const DEFAULT_SPEED = 5.0;
const CALORIES_PER_KM = 55;

class Waypoint {
    constructor(name, lat, lng) {
        this.name = name;
        this.lat = lat;
        this.lng = lng;
    }
}

class RoutePlanner {
    constructor() {
        this.waypoints = [];
        this.speed = DEFAULT_SPEED;
        this.weight = 70;
        this.load();
    }

    load() {
        if (fs.existsSync(DATA_FILE)) {
            try {
                const data = JSON.parse(fs.readFileSync(DATA_FILE));
                this.waypoints = data.waypoints.map(w => new Waypoint(w.name, w.lat, w.lng));
                this.speed = data.speed || DEFAULT_SPEED;
                this.weight = data.weight || 70;
            } catch (e) {}
        }
    }

    save() {
        fs.writeFileSync(DATA_FILE, JSON.stringify({
            waypoints: this.waypoints,
            speed: this.speed,
            weight: this.weight
        }, null, 2));
    }

    async geocode(query) {
        const url = 'https://nominatim.openstreetmap.org/search';
        const params = { q: query, format: 'json', limit: 1 };
        const resp = await axios.get(url, { params, headers: { 'User-Agent': 'WalkRoutePlanner/1.0' }, timeout: 10000 });
        if (resp.data.length === 0) return null;
        return { lat: parseFloat(resp.data[0].lat), lng: parseFloat(resp.data[0].lon) };
    }

    async add(location) {
        if (location.includes(',')) {
            const parts = location.split(',').map(s => s.trim());
            if (parts.length === 2) {
                const lat = parseFloat(parts[0]);
                const lng = parseFloat(parts[1]);
                if (!isNaN(lat) && !isNaN(lng)) {
                    const name = `(${lat.toFixed(4)},${lng.toFixed(4)})`;
                    this.waypoints.push(new Waypoint(name, lat, lng));
                    this.save();
                    console.log(`✅ Added waypoint: ${name}`);
                    return;
                }
            }
        }
        const coords = await this.geocode(location);
        if (!coords) {
            console.log(`❌ Could not geocode '${location}'`);
            return;
        }
        this.waypoints.push(new Waypoint(location, coords.lat, coords.lng));
        this.save();
        console.log(`✅ Added waypoint: ${location} (${coords.lat.toFixed(4)}, ${coords.lng.toFixed(4)})`);
    }

    list() {
        if (this.waypoints.length === 0) {
            console.log('No waypoints.');
            return;
        }
        console.log('\n📍 Waypoints:');
        this.waypoints.forEach((w, i) => {
            console.log(`${i+1}: ${w.name} (${w.lat.toFixed(4)}, ${w.lng.toFixed(4)})`);
        });
    }

    remove(index) {
        if (index < 1 || index > this.waypoints.length) {
            console.log(`❌ Invalid index. Must be between 1 and ${this.waypoints.length}`);
            return;
        }
        const removed = this.waypoints.splice(index-1, 1)[0];
        this.save();
        console.log(`✅ Removed waypoint: ${removed.name}`);
    }

    clear() {
        this.waypoints = [];
        this.save();
        console.log('✅ All waypoints cleared.');
    }

    speed(speed) {
        this.speed = speed;
        this.save();
        console.log(`✅ Walking speed set to ${speed.toFixed(1)} km/h`);
    }

    async route() {
        if (this.waypoints.length < 2) {
            console.log('Need at least 2 waypoints to calculate a route.');
            return;
        }
        const coords = this.waypoints.map(w => `${w.lng},${w.lat}`).join(';');
        const url = `http://router.project-osrm.org/route/v1/walking/${coords}?overview=full&geometries=polyline&steps=true`;
        try {
            const resp = await axios.get(url, { timeout: 10000 });
            const data = resp.data;
            if (!data.routes || data.routes.length === 0) {
                console.log('❌ No route found.');
                return;
            }
            const route = data.routes[0];
            const distanceKm = route.distance / 1000;
            const durationMin = route.duration / 60;
            const elevation = distanceKm * 12;
            const calories = distanceKm * CALORIES_PER_KM * (this.weight / 70);

            console.log('\n🚶 Walking Route:');
            console.log(`Total distance: ${distanceKm.toFixed(2)} km`);
            console.log(`Total duration: ${durationMin.toFixed(1)} min (at ${this.speed.toFixed(1)} km/h)`);
            console.log(`Elevation gain: ~${elevation.toFixed(0)} m`);
            console.log(`Calories burned: ~${calories.toFixed(0)} kcal`);

            if (route.legs) {
                console.log('\nTurn-by-turn:');
                let stepNum = 1;
                for (const leg of route.legs) {
                    for (const step of leg.steps || []) {
                        const instruction = step.maneuver?.instruction || '';
                        const dist = step.distance / 1000;
                        if (instruction) {
                            console.log(`${stepNum}. ${instruction} (${dist.toFixed(2)} km)`);
                            stepNum++;
                        }
                    }
                }
            }
        } catch (err) {
            console.log(`❌ Error fetching route: ${err.message}`);
        }
    }
}

program
    .command('add <location>')
    .action(async (location) => {
        const planner = new RoutePlanner();
        await planner.add(location);
    });

program
    .command('list')
    .action(() => {
        const planner = new RoutePlanner();
        planner.list();
    });

program
    .command('route')
    .action(async () => {
        const planner = new RoutePlanner();
        await planner.route();
    });

program
    .command('remove <index>')
    .action((index) => {
        const planner = new RoutePlanner();
        planner.remove(parseInt(index));
    });

program
    .command('clear')
    .action(() => {
        const planner = new RoutePlanner();
        planner.clear();
    });

program
    .command('speed <kmh>')
    .action((speed) => {
        const planner = new RoutePlanner();
        planner.speed(parseFloat(speed));
    });

program.parse(process.argv);

File: WalkRoute.cs
// WalkRoute.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading.Tasks;

class Waypoint
{
    [JsonPropertyName("name")]
    public string Name { get; set; }
    [JsonPropertyName("lat")]
    public double Lat { get; set; }
    [JsonPropertyName("lng")]
    public double Lng { get; set; }

    public Waypoint() { }
    public Waypoint(string name, double lat, double lng)
    {
        Name = name;
        Lat = lat;
        Lng = lng;
    }
}

class RouteData
{
    [JsonPropertyName("waypoints")]
    public List<Waypoint> Waypoints { get; set; } = new List<Waypoint>();
    [JsonPropertyName("speed")]
    public double Speed { get; set; } = 5.0;
    [JsonPropertyName("weight")]
    public double Weight { get; set; } = 70.0;
}

class RoutePlanner
{
    private RouteData data = new RouteData();
    private readonly string dataFile = "walk_route.json";
    private readonly HttpClient client = new HttpClient();
    private readonly JsonSerializerOptions options = new JsonSerializerOptions { WriteIndented = true };

    public RoutePlanner() => Load();

    private void Load()
    {
        if (!File.Exists(dataFile)) return;
        string json = File.ReadAllText(dataFile);
        data = JsonSerializer.Deserialize<RouteData>(json) ?? new RouteData();
    }

    private void Save()
    {
        string json = JsonSerializer.Serialize(data, options);
        File.WriteAllText(dataFile, json);
    }

    private async Task<(double lat, double lng)?> Geocode(string query)
    {
        string url = $"https://nominatim.openstreetmap.org/search?q={Uri.EscapeDataString(query)}&format=json&limit=1";
        using var req = new HttpRequestMessage(HttpMethod.Get, url);
        req.Headers.Add("User-Agent", "WalkRoutePlanner/1.0");
        var resp = await client.SendAsync(req);
        if (!resp.IsSuccessStatusCode) return null;
        string json = await resp.Content.ReadAsStringAsync();
        var arr = JsonSerializer.Deserialize<List<Dictionary<string, string>>>(json);
        if (arr == null || arr.Count == 0) return null;
        double lat = double.Parse(arr[0]["lat"]);
        double lng = double.Parse(arr[0]["lon"]);
        return (lat, lng);
    }

    public async Task Add(string location)
    {
        if (location.Contains(','))
        {
            var parts = location.Split(',').Select(s => s.Trim()).ToArray();
            if (parts.Length == 2 && double.TryParse(parts[0], out double lat) && double.TryParse(parts[1], out double lng))
            {
                string name = $"({lat:F4},{lng:F4})";
                data.Waypoints.Add(new Waypoint(name, lat, lng));
                Save();
                Console.WriteLine($"✅ Added waypoint: {name}");
                return;
            }
        }
        var coords = await Geocode(location);
        if (!coords.HasValue)
        {
            Console.WriteLine($"❌ Could not geocode '{location}'");
            return;
        }
        data.Waypoints.Add(new Waypoint(location, coords.Value.lat, coords.Value.lng));
        Save();
        Console.WriteLine($"✅ Added waypoint: {location} ({coords.Value.lat:F4}, {coords.Value.lng:F4})");
    }

    public void List()
    {
        if (data.Waypoints.Count == 0)
        {
            Console.WriteLine("No waypoints.");
            return;
        }
        Console.WriteLine("\n📍 Waypoints:");
        for (int i = 0; i < data.Waypoints.Count; i++)
        {
            var w = data.Waypoints[i];
            Console.WriteLine($"{i+1}: {w.Name} ({w.Lat:F4}, {w.Lng:F4})");
        }
    }

    public void Remove(int index)
    {
        if (index < 1 || index > data.Waypoints.Count)
        {
            Console.WriteLine($"❌ Invalid index. Must be between 1 and {data.Waypoints.Count}");
            return;
        }
        var removed = data.Waypoints[index-1];
        data.Waypoints.RemoveAt(index-1);
        Save();
        Console.WriteLine($"✅ Removed waypoint: {removed.Name}");
    }

    public void Clear()
    {
        data.Waypoints.Clear();
        Save();
        Console.WriteLine("✅ All waypoints cleared.");
    }

    public void Speed(double speed)
    {
        data.Speed = speed;
        Save();
        Console.WriteLine($"✅ Walking speed set to {speed:F1} km/h");
    }

    public async Task Route()
    {
        if (data.Waypoints.Count < 2)
        {
            Console.WriteLine("Need at least 2 waypoints to calculate a route.");
            return;
        }
        string coords = string.Join(";", data.Waypoints.Select(w => $"{w.Lng},{w.Lat}"));
        string url = $"http://router.project-osrm.org/route/v1/walking/{coords}?overview=full&geometries=polyline&steps=true";
        var resp = await client.GetAsync(url);
        if (!resp.IsSuccessStatusCode)
        {
            Console.WriteLine("❌ Error fetching route.");
            return;
        }
        string json = await resp.Content.ReadAsStringAsync();
        using var doc = JsonDocument.Parse(json);
        var root = doc.RootElement;
        if (!root.TryGetProperty("routes", out var routesElem) || routesElem.GetArrayLength() == 0)
        {
            Console.WriteLine("❌ No route found.");
            return;
        }
        var route = routesElem[0];
        double distanceKm = route.GetProperty("distance").GetDouble() / 1000;
        double durationMin = route.GetProperty("duration").GetDouble() / 60;
        double elevation = distanceKm * 12;
        double calories = distanceKm * 55 * (data.Weight / 70.0);

        Console.WriteLine("\n🚶 Walking Route:");
        Console.WriteLine($"Total distance: {distanceKm:F2} km");
        Console.WriteLine($"Total duration: {durationMin:F1} min (at {data.Speed:F1} km/h)");
        Console.WriteLine($"Elevation gain: ~{elevation:F0} m");
        Console.WriteLine($"Calories burned: ~{calories:F0} kcal");

        if (route.TryGetProperty("legs", out var legsElem))
        {
            Console.WriteLine("\nTurn-by-turn:");
            int stepNum = 1;
            foreach (var leg in legsElem.EnumerateArray())
            {
                if (leg.TryGetProperty("steps", out var stepsElem))
                {
                    foreach (var step in stepsElem.EnumerateArray())
                    {
                        string instruction = step.GetProperty("maneuver").GetProperty("instruction").GetString();
                        double dist = step.GetProperty("distance").GetDouble() / 1000;
                        Console.WriteLine($"{stepNum}. {instruction} ({dist:F2} km)");
                        stepNum++;
                    }
                }
            }
        }
    }

    static async Task Main(string[] args)
    {
        if (args.Length < 1)
        {
            Console.WriteLine("Usage: WalkRoute <command> [options]");
            return;
        }
        var planner = new RoutePlanner();
        string cmd = args[0];
        switch (cmd)
        {
            case "add":
                if (args.Length < 2) { Console.WriteLine("add <location>"); return; }
                await planner.Add(args[1]);
                break;
            case "list":
                planner.List();
                break;
            case "route":
                await planner.Route();
                break;
            case "remove":
                if (args.Length < 2) { Console.WriteLine("remove <index>"); return; }
                planner.Remove(int.Parse(args[1]));
                break;
            case "clear":
                planner.Clear();
                break;
            case "speed":
                if (args.Length < 2) { Console.WriteLine("speed <kmh>"); return; }
                planner.Speed(double.Parse(args[1]));
                break;
            default:
                Console.WriteLine("Unknown command. Use add, list, route, remove, clear, speed.");
                break;
        }
    }
}
